### PR TITLE
Update to ibm_watson library

### PR DESCRIPTION
**Pull request by acpang**
Replaced 'watson_developer_cloud' with 'ibm_watson':
validate_workspace/validateWS.py
utils/workspaceParser.py
utils/trainConversation.py
utils/get_intent_description.py
utils/__init__.py
run.py
extract_utterances/getUtteranceBeforeNode.py
dialog_test/flowtest_v1.py

Other fixes:
validate_workspace/validateWS.py - getTitle - return Id if Title not available.
utils/get_intent_description.py - corrected JSON reference in write_output method
utils/__init__.py - reduced fold default from 5 to 3 to allow for test_run.py to efficiently operate.

DCO 1.1 Signed-off-by: Andrew Pang apang@us.ibm.com

### CODE REVIEW

**Comment by trangelier** [comment]
I'm not proficient enough in python to really review or provide support to this PR, but I just started making some of the same changes on my local copy then found this.

+1 to @acpang !!

**Comment by ajainghub** [comment]
Hi Andrew,
Due to the recent ibm_watson library changes my Python Notebook that uses this tool's master codebase was also giving problems of type 'unexpected keyword argument' etc. 

I found the fix to be basically all that you have in this commit about using the Authenticator but also need upgraded  ibm-watson==4.0.1 

I tested with you branch but failed when using ibm-watson==3.4.0 (as in the requirements.txt)

`!pip install --upgrade ibm_watson`
Once I was on ibm-watson-4.0.1 everything from your branch works!

In short - Your commits look good but I think you also need to update requirements.txt for 
ibm-watson==4.0.1
Can you check/confirm in your tests which version of ibm-watson you were using?

**Comment by andrewrfreed** [comment]
I am using

```
pip3 list | grep -i ibm
ibm-cloud-sdk-core     1.0.0    
ibm-watson             4.0.1    
```

File: utils/intentmetrics.py [hunk 1]
```diff
@@ -60,6 +60,10 @@ def func(args):
 
             precisions[idx] = precision
             recalls[idx] = recall
+            print ("precision")
+            print (precision)
+            print ("recall")
+            print (recall)
```
Comment (andrewrfreed) [inline utils/intentmetrics.py]: These trace statements should not be printed

File: utils/__init__.py [hunk 2]
```diff
@@ -68,7 +69,7 @@
 BLIND_TEST = 'blind'
 STANDARD_TEST = 'test'
 
-FOLD_NUM_DEFAULT = 5
+FOLD_NUM_DEFAULT = 3
```
Comment (andrewrfreed) [inline utils/__init__.py]: Let's not change this value in this pull request

File: dialog_test/flowtest_v1.py [hunk 3]
```diff
@@ -27,7 +28,12 @@ class FlowTestV1:
                         'Intent9', 'Confidence9', 'Intent10', 'Confidence10' ]
     
     def __init__(self, **kwargs):
-        self.ctk = ConversationV1(**kwargs)
+        authenticator = IAMAuthenticator(kwargs.get("password"))
+        self.ctk = AssistantV1(
+            version=kwargs.get("version"),
+            authenticator=authenticator
+        )
+        self.ctk.set_service_url("https://gateway.watsonplatform.net/assistant/api")
```
Comment (andrewrfreed) [inline dialog_test/flowtest_v1.py]: The url should be passed as one of the arguments to this function, was it not before? (I recall testing this with US-East URL)

File: utils/get_intent_description.py [hunk 4]
```diff
@@ -33,11 +36,13 @@ def add_output_arg(parser):
...
 def get_remote_workspace(args):
+    authenticator = IAMAuthenticator(args.password)
     conv = AssistantV1(
-        WA_API_VERSION,
-        username=args.user,
-        password=args.password
+        version=WA_API_VERSION,
+        authenticator=authenticator
     )
+    conv.set_service_url(BASE_URL)
```
Comment (andrewrfreed) [inline utils/get_intent_description.py]: If this utility did not support a configurable URL, please create an Issue about that - this will not work when the Assistant is NOT located in US-South

File: utils/intentmetrics.py (hunk 1 above)
Comment (acpang) [reply utils/intentmetrics.py]: doh, that wasn't meant to go in.  I'll get rid of it.

File: dialog_test/flowtest_v1.py (hunk 3 above)
Comment (acpang) [reply dialog_test/flowtest_v1.py]: Since the ibm_watson library change the auth method to use IAMAuthenticator, it now requires the url.  So a reference wasn't in this script before.  I'll re-examine the args.

**Comment by andrewrfreed** [comment]
I tested based on commit 3325717 (https://github.com/cognitive-catalyst/WA-Testing-Tool/pull/98/commits/3325717d52a7dfee75df5758f58661b0eba8f5e8) and it worked on all three modes.

I know you are revising further but I'm marking successful test to this point.

File: utils/intentmetrics.py (hunk 1 above)
Comment (acpang) [reply utils/intentmetrics.py]: Reverted change.

File: utils/__init__.py (hunk 2 above)
Comment (acpang) [reply utils/__init__.py]: Changed back to default to 5.

File: dialog_test/flowtest_v1.py (hunk 3 above)
Comment (acpang) [reply dialog_test/flowtest_v1.py]: Passed in the url argument from flowiest.py.

File: utils/get_intent_description.py (hunk 4 above)
Comment (acpang) [reply utils/get_intent_description.py]: Added -l as a required argument for the Watson Assistant Url

File: dialog_test/flowtest.py [hunk 5]
```diff
@@ -3,13 +3,16 @@
 import flowtest_v1
 import pkg_resources
 import pandas as pd
+import validators
 
 DATA_FOLDER='tests'
 OUTPUT_FOLDER='results'
 USERNAME=os.environ["ASSISTANT_USERNAME"]
 PASSWORD=os.environ["ASSISTANT_PASSWORD"]
+WA_URL=os.environ["ASSISTANT_URL"]
```
Comment (andrewrfreed) [inline dialog_test/flowtest.py]: Use WA_URL = os.environ.get("ASSISTANT_URL",url_default)) and you can remove the rest of the changes.  I would prefer not to add another external package (`validators`)

File: dialog_test/flowtest.py [hunk 6]
```diff
@@ -34,6 +37,13 @@ def validateArguments():
         print ('No valid WORKSPACE_ID specified in environment')
         sys.exit(-4)
 
+    if validators.url(WA_URL):
```
Comment (andrewrfreed) [inline dialog_test/flowtest.py]: The warning is a good idea, you can do it without the extra package by using
`if "ASSISTANT_URL" in os.environ`

File: requirements.txt [hunk 7]
```diff
@@ -8,3 +8,4 @@ configparser
 squarify>=0.3.0
 ibm-watson==4.0.1
 seaborn>=0.9.0
+validators==0.14.0
```
Comment (andrewrfreed) [inline requirements.txt]: Prefer to remove this as stated above.
